Persistence

lolmaus opened the pull request:
This PR allows Mirage DB survive page refreshes.

Here's a sketch of documentation page (gonna be filed as a separate PR to the `gh-pages` branch):

> Mirage can be configured for the database to survive page reloads. This is useful for creating a webapp prototype that behaves like a real app but does not require a web server. This prototype can be deployed to a static hosting such as GitHub Pages and offered for everyone to try out without a need to install and compile anything.
> 
> When persistence is enabled, Mirage will dump the database to localStorage after every POST, PUT, PATCH and DELETE request. On app initialization, Mirage loads the dump to the memory.
> 
> 
> 
> ## Enabling persistence
> 
> To enable persistence, set the `persist` option to true in `config/environment.js`:
> 
> ```js
> // config/environment.js
> ...
> if (environment !== 'test') {
>   ENV['ember-cli-mirage'] = {
>     enabled: true,
>     persistence: true,
>   };
> }
> ```
> 
> Note that in `test` environment, persistence will be always disabled even if you enable it explicitly.
> 
> 
> 
> ## Populating the database
> 
> When the app starts and there's no persistence dump, Mirage will run the default scenario from the `mirage/scenarios/default.js` module. See [Using factories in development](../seeding-your-database/#using-factories-in-development) for details.
> 
> When the app starts and the persistence dump is available, it will load the dump *instead* of running the scenario.
> 
> 
> 
> ## Resetting the cache
> 
> To reset the cache manually, you can remove the `ember-cli-mirage-persistence` key from your `localStorage`.
> 
> But there's a larger problem: as keep releasing updates to the scenario or DB schema, users of the prototype will have the database state frozen at different stages. Eventually it will become incompatible with the app, so there's a need to reset the persistence dump for everyone.
> 
> For that purpose, Mirage saves a version with the persistence dump. The version can be any truthy number or string and is `1` by default. It is defined in `config/environment.js`:
> 
> ```js
> // config/environment.js
> ...
> ENV['ember-cli-mirage'] = {
>     enabled: true,
>     persistence: true,
>     persistenceVersion: 1
> };
> ```
> 
> Increment it every time you make a change to the DB schema or update the default scenario.
> 
> When Mirage starts with persistence enabled and finds a dump, it compares the version of the dump with the version defined in `config/environment.js`. If there's a mismatch, Mirage will delete the dump and run the default scenario to create a new one.
> 
> 
> 
> ## Dumping manually
> 
> By default, Mirage will make a database dump after every mutating HTTP request: `POST`, `PUT`, `PATCH` and `DELETE`.
> 
> If you need to also dump after a different request, call `server.savePersistenceDump()` in your route handler.
> 
> 
> 
> ## Persistence and testing
> 
> Enabling persistence would break tests, so Mirage keeps it disabled in the `test` environment.
> 
> If you visit `/tests` in `development` environment, i. e. when running `ember serve`, you'll see tests failing because persistence would load the default scenario and carry app state over from test to test.
> 
> In order to review tests is the browser of a persistence-enabled app, use `ember test --serve` instead of `ember serve`.

***

I'm ready to make changes to this PR according to your suggestions. Here are possible directions:

* Add some tests. Making true acceptance tests is impossible with Ember infrastructure because reloading the page would restart all tests. But I can add some unit tests if you demand.
* Abstract `window.localStorage`. If you demand this, please suggest a way.
* Allow customizing the condition that determines whether the dump should be persisted for a request or not. Currently, it is based on the HTTP verb. I'm not in favor of allowing customization, because:
  * broader API surface
  * more complicated code
  * if you need to, you can simply call `server.savePersistenceDump()` from your route handler.

samselikoff commented:
Amazing work!

samselikoff commented:
> If you visit /tests in development environment, i. e. when running ember serve, you'll see tests failing because persistence would load the default scenario and carry app state over from test to test.

We could check the runtime environment (in addition to build-time) to avoid this

samselikoff commented on addon/server.js:
```diff
@@ -186,7 +193,7 @@ export default class Server {
     } else if (!this.isTest() && hasDefaultScenario) {
       this.loadFactories(config.factories);
-      config.scenarios.default(this);
+      this.loadScenario();
```
Let's call this `populateDatabase` or `seedDatabase` since scenario has a more specific meaning (most folks will think of the `/scenarios` dir).

lolmaus commented:
> We could check the runtime environment (in addition to build-time) to avoid this

How do I do that?

samselikoff commented on addon/server.js:
```diff
@@ -153,6 +158,8 @@ export default class Server {
     this.timing = this.timing || config.timing || 400;
     this.namespace = this.namespace || config.namespace || '';
     this.urlPrefix = this.urlPrefix || config.urlPrefix || '';
+    this.persistence = !this.isTest() && config.addonConfig && config.addonConfig.persistence;
```
Can we change the ENV option to `persistDb` (instead of `persistence`)?

```js
ENV['ember-cli-mirage'] = {
  enabled: true,
  persistDb: true
};
```

feel like it gives a bit more clarity.

We can still keep the `persistence` naming throughout the internal implementation here.

samselikoff commented on addon/server.js:
```diff
@@ -261,6 +287,104 @@ export default class Server {
+    this.log("Mirage persistence dump found, loading it into the database:", data);
+
+    this.db.loadData(data);
+  }
```
I find the early returns in this method hard to read, could we move the validation logic into a method and have no early returns? It would also keep the style more consistent with other parts of the addon.

Something like

```js
loadScenario() {
  let dump = window.localStorage.getItem(LOCALSTORAGE_KEY_DUMP);
  let dumpVersionString = window.localStorage.getItem(LOCALSTORAGE_KEY_VERSION);
  let dumpVersion = (dumpVersion !== undefined) ? JSON.parse(dumpVersion) : dumpVersionString;

  if (this._shouldUseLocalStorageDump()) {
    let data = JSON.parse(dump);
    this.log("Mirage persistence dump found, loading it into the database:", data);
    this.db.loadData(data);
    
  } else {
    this.loadDefaultScenario();
  }
}

_shouldUseLocalStorageDump() {
  let shouldUseDump = true;
  
  if (!this.persistence) {
    shouldUseDump = false;
    
  } else if (!window.localStorage) {
    this.log("Mirage persistence is enabled, but localStorage is unavailable. Loading the default scenario.");
    shouldUseDump = false;
    
  } else if (dumpVersion === undefined && this.persistenceVersion !== undefined) {
    this.log("Mirage persistence dump not found. Loading the default scenario.");
    shouldUseDump = false;
    
  } else if (this.persistenceVersion !== dumpVersion) {
    this.log("Mirage persistence version mismatch. Loading the default scenario.");
    window.localStorage.removeItem(LOCALSTORAGE_KEY_VERSION);
    window.localStorage.removeItem(LOCALSTORAGE_KEY_DUMP);
    shouldUseDump = false;
    
  } else if (!dump) {
    this.log("Mirage persistence dump not found. Loading the default scenario.");
    window.localStorage.removeItem(LOCALSTORAGE_KEY_VERSION);
    shouldUseDump = false;
  }

  return shouldUseDump;
}
```

samselikoff commented on addon/server.js:
```diff
@@ -474,7 +598,8 @@ export default class Server {
       verb, rawHandler, customizedCode, options, path,
-      serializerOrRegistry: this.serializerOrRegistry
+      serializerOrRegistry: this.serializerOrRegistry,
+      server: this
```
why are we passing `server` to RouteHandler now?

lolmaus replied:
Because we need to trigger persistence after a mutating request.

samselikoff replied:
I'm still not seeing it... doesn't `server.js` persist the db by calling `savePersistenceDump()` from the `handledRequest` hook?

lolmaus replied:
Reverted.

samselikoff commented:
>> We could check the runtime environment (in addition to build-time) to avoid this

> How do I do that?

It comes from

```
import ENV from '../config/environment';
```

but it should already be a property on the `server` instance, `server.environment`. If that's `test` you know you're in testing land.

lolmaus commented:
> but it should already be a property on the server instance, server.environment. If that's test you know you're in testing land.

So `server.isTest()` is true when visiting `/tests` in `development` environment?

samselikoff commented:
One more thing, I know it seems like we should disable during testing but, I'm wondering if we can just set that as the default, but if people explicitly enable it then it will work for them.

We do this with the default value of `enabled`, which changes if the user is using `--proxy`. But, if they ever explicitly set the `enabled` property in their `config`, that value always wins.

What do you think about this?

samselikoff commented:
> So server.isTest() is true when visiting /tests in development environment?

should be, yes. Believe that's the runtime environment (and we set it up like this for precisely the same reason - need to know if someone is in a build-time of `development` but run-time of `test`.

(By the way - it is better for folks to do `ember test -s` but I think some people still visit `/tests` so it needs to be supported, for now.)

lolmaus commented:
> One more thing, I know it seems like we should disable during testing but, I'm wondering if we can just set that as the default, but if people explicitly enable it then it will work for them.
> 
> We do this with the default value of enabled, which changes if the user is using --proxy. But, if they ever explicitly set the enabled property in their config, that value always wins.
>
> What do you think about this?

It's already possible to override `sever.persistence`, enabling persistence unconditionally.

I didn't document it because I failed to find a single use case for this. It will certainly break any tests because app state will leak through them.

Can you think of at least one legit use case?

lolmaus commented:
>> So server.isTest() is true when visiting /tests in development environment?
> 
> should be, yes. Believe that's the runtime environment (and we set it up like this for precisely the same reason - need to know if someone is in a build-time of development but run-time of test.

I'm already using `server.isTest()`, so it should already work. I'll remove the warning from documentation.

samselikoff commented:
> Can you think of at least one legit use case?

well we could write one or two acceptance tests for this PR ;)

samselikoff commented:
> It's already possible to override sever.persistence, enabling persistence unconditionally.

I think the current implementation

```
this.persistence = !this.isTest() && config.addonConfig && config.addonConfig.persistence;
```

makes it impossible to enable persistence during testing. right?

samselikoff commented:
I think it'd be good to have one acceptance test that sets some data in `localStorage`, and then asserts that it ended up in `server.db`.

By the way I think we might have a test where we override the Server's environment, and make it "act" like a `development` server. Maybe this is all we need to test.

lolmaus commented:
> makes it impossible to enable persistence during testing. right?

Can't we simply do `server.persistence = true` in a test?

lolmaus commented:
@samselikoff I've made changes according to your comments and added an acceptance test.

samselikoff commented:
Thank you for your changes.

- [ ] "Can't we simply do server.persistence = true in a test?" — actually you're right, I think this is better, because if somebody adds `persistDb: true` in their `config` file, that by default would explicitly enable it for all environments. So we'll disable it in testing unless they override it on the server instance, just like we do for `logging`. 👍  this makes sense.

- [ ] I'd rather see the test verify the outer-most wiring & functionality, and since users won't be calling `localStorage.setItem` or `server.populateDatabase()` let's try something like this.

```js
import {test} from 'qunit';
import moduleForAcceptance from '../helpers/module-for-acceptance';

moduleForAcceptance('Acceptance | Persistence', {
  before() {
    server.resource('user');
  }
});

test('new data is persisted to localStorage', async function(assert) {
  server.persistDb = true;
  
  await $.ajax({
    method: 'POST',
    url: '/users',
    contentType: 'application/json',
    data: JSON.stringify({
      data: {
        type: 'users',
        attributes: {
          'first-name': 'Andrey'
        }
      }
    })
  });

  let data = window.localStorage.getItem('ember-cli-mirage:persistence');
  // Assert Andrey is in data
  assert.equal(...);
});

test('existing data is read from localStorage', async function(assert) {
  window.localStorage.setItem(...);
  server.persistDb = true;
  
  let results = await $.ajax({
    method: 'GET',
    url: '/users/1',
  });

  assert.equal(results, ...);
});
```

samselikoff commented:
Also for the tests I wrote I realize `server` will already be initialized before we get to `server.persistDb` line, so maybe you should look at integration/server tests and write these there. Then you can just instantiate a server, and set the localStorage data before it & pass in the appropriate config.

lolmaus commented:
The test that you want requires some changes:

1. [This logic](https://github.com/lolmaus/ember-cli-mirage/blob/631888f6b84cf5bab0213ca26b0992a379d0c856/addon/server.js#L192-L199) currently prevents scenarios from being loaded in tests. We have to make it even more test-aware. I didn't do it initially because if your app has special cases for a test, it's not a real test.
2. Overriding server options in such a way:

    ```js
    let server = new Server();
    server.persistDb = true;
    ```

    ...is not possible because the option is evaluated in the constructor, i. e. before it's overridden.

    The logic of overriding option values has to be updated, so that options can be passed like this instead:

    ```js
    let server = new Server({
      persistDb: true
    });
    ```

I've filed a PR to my PR with these changes: https://github.com/lolmaus/ember-cli-mirage/pull/2/files

When I implemented a new way of applying default and override options, the `loadConfig` method became redundant, so I removed it. BTW, it had incorrect inline documentation.

I also removed the `baseConfig` option because it was not used anywhere, never tested and never documented.

Please tell me if you want to have these changes merged into this PR.

samselikoff commented:
There's a lot going on here. The amount of code changes makes it harder to move this PR forward. I really appreciate the work - in the future let's try to keep everything small and focused, so it makes it easier to review.

1. `loadConfig` is public so some people rely on it, therefore we can't remove it. If the docs are broken ({@Object} declaration for `config` instead of function, I believe you're referring to) let's fix it.

2. Do you mind removing `_applyOptionsWithDefaults`, I think it is decent but unnecessary, as other devs will need to read that method to understand what's going on. Setting instance variables in the constructor is common enough.

3. The changes to passthrough-test are fine, I think those were written earlier so I like this style.

4. For your second PR's test, can we
  - take away `environment: test`, `timing: 0`, `persistDbVersion: 1` (since these would not be set by a user, and the goal here is to mimic real-world use)
  - move the `const data = {}` and `localStorage.setItem` code to inside of the "should respond with data from localStorage" test
  - add a test for "should persist data to localStorage" and check that a call does so?
  - also the `server.get` route definition can use the shorthand `server.get('/contacts')`.

5. Let's add back in the `loadConfig` unit tests

caseywatts commented:
This is awesome! I can't wait for this feature 🎉 

Is this an issue we could break up into smaller sub-issues and ask the community to work on? I'm happy to try to help break it up :)

@lolmaus @samselikoff

techdragon commented:
Its kind of strange that this issue has been idle so long, is there any progress? Or an alternative solution?

samselikoff commented:
Right now I’m focused on shipping 1.0. I think the alternative will be running Mirage in Node. I think that’s a better solution for several reasons. 

Sent from my iPhone

> On Mar 27, 2019, at 9:59 PM, Samuel Bishop <notifications@github.com> wrote:
> 
> Its kind of strange that this issue has been idle so long, is there any progress? Or an alternative solution?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

samselikoff commented:
Closing this as stale, but keeping a reference to it in https://github.com/miragejs/miragejs/issues/217 in case we want to reference these APIs.
